<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">

<head>

    <meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
    <title>g2gtools documentation</title>
    <meta name="author" content='Kwangbom \"KB\" Choi'/>
    <meta name="description" content="g2gtools: A better tool for alignment target creation and liftover."/>
    <meta name="keywords" content="g2gtools,NGS,personal genome,alignment,target,liftover"/>
    <meta name="Resource-type" content="Document"/>

    <link rel="stylesheet" type="text/css" href="stylesheets/alvaro1.css"/>
    <link rel="stylesheet" type="text/css" href="stylesheets/alvaro2.css"/>
    <link rel="stylesheet" type="text/css" href="stylesheets/alvaro3.css"/>
    <link rel="stylesheet" type="text/css" href="stylesheets/examples.css"/>
    <link rel="stylesheet" href="http://cdnjs.cloudflare.com/ajax/libs/highlight.js/8.7/styles/default.min.css">

    <!--<style>-->

        <!--/* Sections-->
         <!--* -&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45;&#45; */-->
        <!--#section0 img,-->
        <!--#section1 img {-->
            <!--margin: 20px 0 0 0;-->
        <!--}-->

        <!--#section2 img {-->
            <!--margin: 20px 0 0 52px;-->
        <!--}-->

        <!--#section3 img {-->
            <!--bottom: 0px;-->
            <!--position: absolute;-->
            <!--margin-left: -420px;-->
        <!--}-->

        <!--.intro p {-->
            <!--width: 50%;-->
            <!--margin: 0 auto;-->
            <!--font-size: 1.5em;-->
        <!--}-->

        <!--.twitter-share-button {-->
            <!--position: absolute;-->
            <!--z-index: 99;-->
            <!--right: 149px;-->
            <!--top: 9px;-->
        <!--}-->

    <!--</style>-->

    <!--[if IE]>
    <script type="text/javascript">
        var console = {
            log: function () {
            }
        };
    </script>
    <![endif]-->

    <script src="http://ajax.googleapis.com/ajax/libs/jquery/1.8.3/jquery.min.js"></script>
    <script src="http://ajax.googleapis.com/ajax/libs/jqueryui/1.9.1/jquery-ui.min.js"></script>
    <script src="http://cdnjs.cloudflare.com/ajax/libs/highlight.js/8.7/highlight.min.js"></script>
    <script type="text/javascript" src="js/jquery.fullPage.js"></script>

</head>


<body>

<!--<select id="demosMenu">-->
    <!--<option selected>Choose Demo</option>-->
    <!--<option id="backgrounds">Background images</option>-->
    <!--<option id="backgroundVideo">Background video</option>-->
    <!--<option id="gradientBackgrounds">Gradient backgrounds</option>-->
    <!--<option id="backgroundsFixed">Fixed fullscreen backgrounds</option>-->
    <!--<option id="looping">Looping</option>-->
    <!--<option id="noAnchor">No anchor links</option>-->
    <!--<option id="scrollingSpeed">Scrolling speed</option>-->
    <!--<option id="easing">Easing</option>-->
    <!--<option id="callbacks">Callbacks</option>-->
    <!--<option id="css3">CSS3</option>-->
    <!--<option id="continuous">Continuous scrolling</option>-->
    <!--<option id="normalScroll">Normal scrolling</option>-->
    <!--<option id="scrollBar">Scroll bar enabled</option>-->
    <!--<option id="scrolling">Scroll inside sections and slides</option>-->
    <!--<option id="navigationV">Vertical navigation dots</option>-->
    <!--<option id="navigationH">Horizontal navigation dots</option>-->
    <!--<option id="fixedHeaders">Fixed headers</option>-->
    <!--<option id="apple">Apple iPhone demo (animations)</option>-->
    <!--<option id="oneSection">One single section</option>-->
    <!--<option id="responsiveHeight">Responsive Height</option>-->
    <!--<option id="responsiveWidth">Responsive Width</option>-->
    <!--<option id="methods">Methods</option>-->
<!--</select>-->

<ul id="menu">
    <li data-menuanchor="overview"><a href="#overview">Overview</a></li>
    <li data-menuanchor="installation"><a href="#installation">Installation</a></li>
    <li data-menuanchor="features"><a href="#features">Features</a></li>
    <li data-menuanchor="workflow"><a href="#workflow">Workflow</a></li>
</ul>

<div id="fullpage">

    <div class="section" id="section0">
        <div class="intro">
            <h1>g2gtools</h1>
            <p><br>
                <strong>g2gtools</strong> is a new suite of tools that creates personal diploid genomes; extracts regions from
                personal genomes; and lifts over coordinates between the reference and personal genomes. Unlike other liftover
                tools, <strong>g2gtools</strong> does not throw away entries that contain indels.
            </p>
            <br><br>
                <a href="https://github.com/churchill-lab/g2gtools" class="btn">View on GitHub</a>
                <a href="https://github.com/churchill-lab/g2gtools/zipball/master" class="btn">Download .zip</a>
                <a href="https://github.com/churchill-lab/g2gtools/tarball/master" class="btn">Download .tar.gz</a>
        </div>
    </div>

    <div class="section" id="section1">
        <div class="intro">
            <h1>Installation</h1>
            <p>
                We highly recommend using <a href="https://www.continuum.io/downloads">Anaconda distribution</a> of
                python to install all the dependencies without issues, although g2gtools is also available at PyPI for
                ‘pip install’ or ‘easy_install’. The most recent version is also available on Anaconda
                Cloud, so add the following channels if you have not already.
            </p>
            <br>
            <div style="text-align:center">
                <div id="code" style="display:inline-block; text-align:left">
                    <pre><code class="console" style="font-size: 16px;">
$ conda config --add channels r
$ conda config --add channels bioconda
                    </code></pre>
                </div>
            </div>
            <br><br>
            <p>
                To avoid conflicts among dependencies, we highly recommend using conda virtual environment:
            </p>
            <br>
            <div style="text-align:center">
                <div id="code" style="display:inline-block; text-align:left">
                    <pre><code class="console" style="font-size: 16px;">
$ conda create -n g2gtools jupyter ipykernel
$ source activate g2gtools
                    </code></pre>
                </div>
            </div>
            <br><br>
            <p>
                Once g2gtools virtual environment is created and activated, your shell prompt will show ‘(g2gtools)’
                at the beginning to specify what virtual environment you are currently in. Now type the following
                and install g2gtools:
            </p>
            <br>
            <div style="text-align:center">
                <div id="code" style="display:inline-block; text-align:left">
                    <pre><code class="console" style="font-size: 16px;">
(g2gtools) $ conda install -c kbchoi g2gtools
                    </code></pre>
                </div>
            </div>
            <br><br>
            <p>
                That’s all! You can go out from g2gtools virtual environment anytime by deactivating it:
            </p>
            <br>
            <div style="text-align:center">
                <div id="code" style="display:inline-block; text-align:left">
                    <pre><code class="console" style="font-size: 16px;">
(g2gtools) $ source deactivate
                    </code></pre>
                </div>
            </div>
        </div>
    </div>

    <div class="section" id="section2">
        <div class="slide">
            <div class="intro">
                <h1>Features</h1>
                <p>In this section, we will show core features of g2gtools.</p>
                <!--<img src="imgs/slider.png" alt="slider"/>-->
            </div>
        </div>

        <div class="slide">
            <div class="intro">
                <img src="imgs/circled_numbers.01.png" alt="Compatible" width="256"/>
                <h1>vcf2vci</h1>
                <p>
                    compiles variant call information from a VCF file into a G2G VCI format file for later
                    (sample-specific) lookup. The input VCF file should contain both snps and indels.
                </p>
                <br><br><br><br>
                <div style="text-align:center">
                    <div id="code" style="display:inline-block; text-align:left">
                        <pre><code class="console" style="font-size: 16px;">
$ g2gtools vcf2vci -h

Creates VCI file from VCF file(s)

Usage: g2gtools vcf2vci [-i &lt;indel VCF file&gt;]* -s &lt;strain&gt; -o &lt;output file&gt; [options]

Required Parameters:
    -i, --vcf &lt;vcf_file&gt;             VCF file name
    -s, --strain &lt;Strain&gt;            Name of strain (column in VCF file)
    -o, --output &lt;Output file&gt;       VCI file name to create

Optional Parameters:
    -p, --num-processes &lt;number&gt;     The number of processes to use, defaults to the number of cores
    --diploid                        Create diploid VCI
    --keep                           Keep track of VCF lines that could not be converted to VCI file
    --pass                           Use only VCF lines that have a PASS for the filter value
    --quality                        Filter on quality, FI=PASS
    --no-bgzip                       DO NOT compress and index output

Help Parameters:
    -h, --help                       Print the help and exit
    -d, --debug                      Turn debugging mode on (list multiple times for more messages)
                        </code></pre>
                    </div>
                </div>
            </div>
        </div>

        <div class="slide">
            <div class="intro">
                <img src="imgs/circled_numbers.02.png" alt="Compatible" width="256"/>
                <h1>patch</h1>
                <p>
                    replaces bases in the input (reference) sequence using SNPs in the sample-specific VCI file. This
                    command supports patching of local subsequence on the fly.
                </p>
                <br><br><br><br>
                <div style="text-align:center">
                    <div id="code" style="display:inline-block; text-align:left">
                        <pre><code class="console" style="font-size: 16px;">
$ g2gtools patch

Patch SNPs onto the reference sequence

Usage: g2gtools patch -i &lt;Fasta file&gt; -c &lt;VCI file&gt; [options]

Required Parameters:
    -i, --input &lt;Fasta file&gt;         Reference fasta file to extract and patch on
    -c, --vci &lt;VCI file&gt;             VCI file

Optional Parameters:
    -p, --num-processes &lt;number&gt;     The number of processes to use, defaults to the number of cores
    -o, --output &lt;Output file&gt;       Name of output fasta file that contains SNP-patched sequences
    -r, --region &lt;seqid:start-end&gt;   Region to extract (cannot be used with -b)
    -b, --bed &lt;BED file&gt;             BED file (cannot be used with -r)
    --bgzip                          Compress and index output (can only be used with -o)

Help Parameters:
    -h, --help                       Print the help and exit
    -d, --debug                      Turn debugging mode on (list multiple times for more messages)

Note:
    --bgzip can be potentially slow depending on the size of the file
                        </code></pre>
                    </div>
                </div>
            </div>
        </div>

        <div class="slide">
            <div class="intro">
                <img src="imgs/circled_numbers.03.png" alt="Compatible" width="256"/>
                <h1>transform</h1>
                <p>
                    incorporates indels specified in a sample's VCI file into the input (reference) sequence. This
                    command also supports the transformation of local subsequence on the fly.
                </p>
                <br><br><br><br>
                <div style="text-align:center">
                    <div id="code" style="display:inline-block; text-align:left">
                        <pre><code class="console" style="font-size: 16px;">
$ g2gtools transform -h

Incorporate indels onto the input sequence

Usage: g2gtools transform -i &lt;Fasta file&gt; -c &lt;VCI file&gt; [options]

Required Parameters:
    -i, --input &lt;Fasta file&gt;         Reference fasta file to extract and transform
    -c, --vci &lt;VCI file&gt;             VCI file

Optional Parameters:
    -p, --num-processes &lt;number&gt;     The number of processes to use, defaults to the number of cores
    -o, --output &lt;Output file&gt;       Name of output fasta file that contains indel-incorporated sequences
    -r, --region &lt;seqid:start-end&gt;   Region to extract (cannot be used with -b)
    -b, --bed &lt;BED file&gt;             BED file (cannot be used with -r)
    --bgzip                          Compress and index output fasta file (can only be used with -o)

Help Parameters:
    -h, --help                       Print the help and exit
    -d, --debug                      Turn debugging mode on (list multiple times for more messages)

Note:
    --bgzip can be potentially slow depending on the size of the file
                        </code></pre>
                    </div>
                </div>
            </div>
        </div>

        <div class="slide">
            <div class="intro">
                <img src="imgs/circled_numbers.04.png" alt="Compatible" width="256"/>
                <h1>convert</h1>
                <p>
                    lifts over coordinates back and forth between reference and custom sample-specific genomes. It
                    supports BAM/SAM, GTF, and BED formats. We implemented g2gtools since all available liftover tools
                    discard alignments that land on indel locations in BAM/SAM files with no justifiable reason. We
                    wanted to avoid removing such alignments since they are possibly enriched with allele-specific
                    information.
                </p>
                <br><br><br><br>
                <div style="text-align:center">
                    <div id="code" style="display:inline-block; text-align:left">
                        <pre><code class="console" style="font-size: 16px;">
$ g2gtools convert -h

Liftover coordinates of bam|sam|gtf|bed files

Usage: g2gtools convert -c &lt;VCI file&gt; -i &lt;input file&gt; [options]

Required Parameters:
    -i, --input &lt;input file&gt;         Input file to liftover
    -c, --vci &lt;VCI file&gt;             VCI file

Optional Parameters:
    -f, --format &lt;bam|sam|gtf|bed&gt;   Input file format
    -o, --output &lt;output file&gt;       Name of output file
    --reverse                        Reverse the Chain file

Help Parameters:
    -h, --help                       Print the help and exit
    -d, --debug                      Turn debugging mode on (list multiple times for more messages)


Note:
    If no output file is specified [-o], the converted information will be redirected
    to standard out.
                        </code></pre>
                    </div>
                </div>
           </div>
        </div>

        <div class="slide">
            <div class="intro">
                <img src="imgs/circled_numbers.05.png" alt="Compatible" width="256"/>
                <h1>gtf2db</h1>
                <p>
                    creates G2G DataBase (DB) file using gene annotation (GTF format). G2G DB file is used for 'extract'ing
                    annotated elements -- generally exons, transcripts, and gene regions -- from the input genomes.
                </p>
                <br><br><br><br>
                <div style="text-align:center">
                    <div id="code" style="display:inline-block; text-align:left">
                        <pre><code class="console" style="font-size: 16px;">
$ g2gtools gtf2db -h

Convert a GTF file to a G2G DB file

Usage: g2gtools gtf2db -i &lt;GTF file&gt; -o &lt;G2G DB file&gt; [options]

Required Parameters:
    -i, --input &lt;GTF file&gt;           GTF file to parse
    -o, --output &lt;G2G DB file&gt;       G2G DB file to create

Help Parameters:
    -h, --help                       Print the help and exit
    -d, --debug                      Turn debugging mode on (list multiple times for more messages)
                        </code></pre>
                    </div>
                </div>
           </div>
        </div>
        <div class="slide">
            <div class="intro">
                <img src="imgs/circled_numbers.06.png" alt="Compatible" width="256"/>
                <h1>extract</h1>
                <p>
                    excerpts genomic regions from the input genome. Useful for creating FASTA file of exons,
                    transcripts, and genes when used with G2G DB (gene annotation) file. We can specify putative peak
                    regions for ChIP-Seq or ATAC-Seq in a BED format file to create "peakome".
                </p>
                <br><br><br><br>
                <div style="text-align:center">
                    <div id="code" style="display:inline-block; text-align:left">
                        <pre><code class="console" style="font-size: 16px;">
$ g2gtools extract -h

Extract subsequence from a fasta file given a region

Usage: g2gtools extract -i &lt;Fasta file&gt; [-r &lt;region&gt; | -b &lt;BED file&gt; | -db &lt;Database&gt; | -id &lt;seqid&gt;] [options]

Required Parameters:
    -i, --input &lt;Fasta file&gt;         Fasta file to extract subsequence from, BGZIP files supported
    -b, --bed &lt;BED file&gt;             BED file -OR-
    -id, --identifier &lt;identifier&gt;   Fasta identifier -OR-
    -r, --region &lt;seqid:start-end&gt;   Region to extract -OR-
    -db, --database &lt;DB file&gt;        Database file
         --transcripts                 For use with -db, extracts transcripts (default)
         --exons                       For use with -db, extracts exons
         --genes                       For use with -db, extracts genes

Optional Parameters:
    -c, --vci &lt;VCI file&gt;             Input VCI file, matching input Fasta file
    -R, --vci-reverse                Reverse the direction of liftover
    --complement                     Complement the extracted sequence
    --reverse                        Reverse the extracted sequence
    --reverse-complement             Reverse complement the extracted sequence
    --raw                            Just shows the extracted sequence

Help Parameters:
    -h, --help                       Print the help and exit
    -d, --debug                      Turn debugging mode on (list multiple times for more messages)

Note:
    Locations specified on command line are 1-based coordinates.
    Locations specified via BED file are 0-based coordinates.
                        </code></pre>
                    </div>
                </div>
            </div>
        </div>
    </div>


    <div class="section" id="section3">
        <div class="intro">
            <h1>Workflow</h1>
            <p>
                We show a typical workflow for creating diploid genome, exome, and transcriptome using a human example:
                NA19670 from 1000 Genomes (phase 3).
            </p>
            <br><br><br><br>
            <ol>
            <li>1. Create VCI file of the sample</li>
            <div style="text-align:center">
                <div id="code" style="display:inline-block; text-align:left">
                    <pre><code class="console" style="font-size: 16px;">
$ g2gtools vcf2vci -o NA19670.vci -s NA19670 --diploid -p 8 \
-i ALL.chr1.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chr2.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chr3.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chr4.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chr5.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chr6.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chr7.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chr8.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chr9.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chr10.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chr11.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chr12.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chr13.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chr14.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chr15.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chr16.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chr17.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chr18.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chr19.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chr20.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chr21.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chr22.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz \
-i ALL.chrX.phase3_shapeit2_mvncall_integrated_v1b.20130502.genotypes.vcf.gz \
-i ALL.chrY.phase3_integrated_v1b.20130502.genotypes.vcf.gz \
-i ALL.chrMT.phase3_callmom.20130502.genotypes.vcf.gz
                    </code></pre>
                </div>
            </div>
            <br>
            <li>2. Incorporate SNPs into the reference genome.</li>
            <div style="text-align:center">
                <div id="code" style="display:inline-block; text-align:left">
                    <pre><code class="console" style="font-size: 16px;">
$ g2gtools patch -i hs37d5.fa -c NA19670.vci.gz -o NA19670.fa -p 8
                    </code></pre>
                </div>
            </div>
            <br>
            <li>3. Incorporate indels into the reference genome and get custom diploid genome for the sample.</li>
            <div style="text-align:center">
                <div id="code" style="display:inline-block; text-align:left">
                    <pre><code class="console" style="font-size: 16px;">
$ g2gtools transform -i NA19670.patched.fa -c NA19670.vci.gz -o NA19670.fa -p 8
                    </code></pre>
                </div>
            </div>
            <br>
            <li>4. Liftover gene annotation onto sample coordinates.</li>
            <div style="text-align:center">
                <div id="code" style="display:inline-block; text-align:left">
                    <pre><code class="console" style="font-size: 16px;">
$ g2gtools convert -i Homo_sapiens.GRCh37.75.gtf -c NA19670.vci.gz -o NA19670.gtf
                    </code></pre>
                </div>
            </div>
            <br>
            <li>5. Parse custom gene annotation into a G2G database file.</li>
            <div style="text-align:center">
                <div id="code" style="display:inline-block; text-align:left">
                    <pre><code class="console" style="font-size: 16px;">
$ g2gtools gtf2db -i NA19670.gtf -o NA19670.db
                    </code></pre>
                </div>
            </div>
            <br>
            <li>6. Extract exons, transcripts, and gene regions from the custom sample genome.</li>
            <div style="text-align:center">
                <div id="code" style="display:inline-block; text-align:left">
                    <pre><code class="console" style="font-size: 16px;">
$ g2gtools extract -i NA19670.fa -db NA19670.db --exons > NA19670.exons.fa
$ g2gtools extract -i NA19670.fa -db NA19670.db --transcripts > NA19670.transcripts.fa
$ g2gtools extract -i NA19670.fa -db NA19670.db --genes > NA19670.genes.fa
                    </code></pre>
                </div>
            </div>
            </ol>
        </div>
        <!--<img src="imgs/tablets.png" alt="tablets"/>-->
    </div>

</div>

<script type="text/javascript">
    $(document).ready(function () {

        $('#fullpage').fullpage({
            sectionsColor: ['whitesmoke', 'whitesmoke', 'whitesmoke', 'whitesmoke'],
            anchors: ['overview', 'installation', 'features', 'workflow'],
            menu: '#menu',
            scrollingSpeed: 500,
            scrollOverflow: true
        });

        $('pre code').each(function (i, block) {
            hljs.highlightBlock(block);
            console.log(block);
        });
    });
</script>

</body>

</html>


